### ISSUE TITLE
Terminating Gradle build does not execute finalizedBy task

### ISSUE THREAD

**Issue by amimas** [gradle/gradle#13212]
I'm experiencing an issue that has been discussed in the following forum thread:

https://discuss.gradle.org/t/ctrl-c-during-running-build-also-kills-daemon/10158

Based on the above thread, the issue was fixed in version 3.2. I have been using version 5.2 and it has been working so far. I've recently upgraded to Gradle 6.4.1 and seeing the same issue described in that thread.

Below is a snippet of my tasks and their configuration.

```gradle
task startDevServer(type: YarnTask) {
    dependsOn 'yarn_install', 'startPlantUMLContainer'
    finalizedBy 'stopPlantUMLContainer'

    args = ['start']
    environment = [ "PLANTUML_SERVICE": "http://localhost:$plantuml_server_port/"]
}

task startPlantUMLContainer(type: DockerStartContainer) {
  dependsOn createPlantUMLContainer
  targetContainerId createPlantUMLContainer.getContainerId()
}

task stopPlantUMLContainer(type: DockerStopContainer) {
  targetContainerId createPlantUMLContainer.getContainerId()
}
```

Typical workflow is to run `./gradlew startDevServer`, which will start a PlantUML docker container first. Then `startDevServer` stays running in the foreground with a dev server available. To stop the dev server, the gradle build is terminated by `ctrl + c`, which would also run the `stopPlantUMLContainer` task as it was set by `finalizedBy`.

### Expected Behavior

The `stopPlantUMLContainer` task is executed when the gradle build is terminated while running `startDevServer` task.

### Current Behavior

After upgrading from Gradle 5.2 to 6.4.1, the `stopPlantUMLContainer` task is not executed when `startDevServer` task is terminated.

### Context

I needed to upgrade to newer version of Gradle because version 5.2 was not compatible with newer version of Java. Upgrading to latest version addressed this issue, but seem to have ran into a new issue, that wasn't there.

Because the task defined in `finalizedBy` is not being executed when the Gradle build is terminated, the next time I go through the usual steps explained above doesn't work.

### Steps to Reproduce 

Please let me know if the above explanation doesn't help and a self-contained repository is needed. I will look into it.

### Your Environment

```
------------------------------------------------------------
Gradle 6.4.1
------------------------------------------------------------

Build time:   2020-05-15 19:43:40 UTC
Revision:     1a04183c502614b5c80e33d603074e0b4a2777c5

Kotlin:       1.3.71
Groovy:       2.5.10
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          1.8.0_144 (Oracle Corporation 25.144-b01)
OS:           Mac OS X 10.15.3 x86_64
```

**Comment by antanas-arvasevicius** [comment]
I've the same issue with finalizedBy task, finalized task is not called if task is cancelled.
Using `Gradle 6.8.3`

More details: 
I'm using docker compose gradle plugin and have configured spring bootRun task to depend on dockerComposeUp task and finalizedBy with dockerComposeDown task.  
If I'm terminating a bootRun task the dockerComposeDown task is never been called.

**Comment by phil-lavin** [comment]
:+1:

**Comment by kota65535** [comment]
Still an issue on 8.6. Is this an intended behavior?

**Comment by doctorpangloss** [comment]
One solution is to implement Job Objects on Windows. This is how other applications and libraries recreate the behavior of Linux / macOS. The aesthetics of the solution are not very good.

This was documented here: https://github.com/gradle/gradle/issues/18716

Duplicates were: https://github.com/gradle/gradle/issues/7603 https://github.com/gradle/gradle/issues/7603 https://github.com/gradle/gradle/issues/6114 https://github.com/gradle/gradle/issues/3093 https://github.com/gradle/gradle/issues/1128 https://github.com/gradle/gradle/issues/1109

I understand there's a lot of inertia around this issue, and a lot of confusion from commenters about this issue. My issue was marked as a duplicate. Commenters say it's a multi-platform issue.

Kubernetes, the Spotify client, etc., they all do this. It's tough. I really hope there is a solution.

**Comment by CLOVIS-AI** [comment]
The comments here seem to mention this being a problem mainly on Windows, but I can see it as well on Linux.

**Comment by doctorpangloss** [comment]
> The comments here seem to mention this being a problem mainly on Windows, but I can see it as well on Linux.

This is true but they are distinct issues. The specific thing the user is talking about causes two problems on Windows and one on Linux:

| Issue                           | Windows          | Linux            | Solution     |
|---------------------------------|------------------|------------------|--------------|
| Docker doesn't stop             | :white_check_mark: | :white_check_mark: | Ryuk         |
| Interrupt doesn't kill processes | :white_check_mark: | :x:              | Job Objects  |

`finalizedBy` is a bit of a red herring. There's no such thing as interrupting a process and getting graceful shutdown in a general case. That's what Ryuk is meant to address. The user should use Testcontainers for long-running tasks, and should not use Gradle to replace Testcontainers.

This is a journey I went years ago and tried communicating to the Gradle and the developer of NuProcess. It's tough. I just use a fork. There's a lot of inertia around code that already exists. They interpret people writing cranky comments as a personality flaw, instead of a measure of how serious the issue is. Time will tell if this will ever be addressed.

**Comment by ov7a** [comment]
Sorry for the late reply.

The issue is in the backlog of the relevant team and is prioritized by them.

***

As indicated in the comments, this is a complex topic. Finding a solution will require a deeper investigation from the responsible team.

Related:
* https://github.com/gradle/gradle/issues/7603
* https://github.com/gradle/gradle/pull/18756
